diesel/src/connection/transaction_manager.rs

```diff
@@ -117,31 +117,120 @@ where
+            }
+        } else {
+            self.change_transaction_depth(
+                -1,
                 conn.batch_execute(&format!(
                     "RELEASE SAVEPOINT diesel_savepoint_{}",
                     transaction_depth - 1
-                ))
-            },
-        )
+                )),
+            )
+        }
     }
 
     fn get_transaction_depth(&self) -> u32 {
         self.transaction_depth.get() as u32
     }
 }
+
+#[cfg(test)]
+mod test {
+
+    #[test]
+    #[cfg(feature = "postgres")]
+    fn transaction_depth_is_tracked_properly_on_commit_failure() {
+        use crate::result::DatabaseErrorKind::SerializationFailure;
+        use crate::result::Error::DatabaseError;
+        use crate::*;
+        use assert_matches::assert_matches;
+        use std::sync::{Arc, Barrier};
+        use std::thread;
+
+        table! {
+            #[sql_name = "transaction_depth_is_tracked_properly_on_commit_failure"]
+            serialization_example {
+                id -> Serial,
+                class -> Integer,
+            }
+        }
+
+        let conn = crate::test_helpers::pg_connection_no_transaction();
+
+        sql_query("DROP TABLE IF EXISTS transaction_depth_is_tracked_properly_on_commit_failure;")
+            .execute(&conn)
+            .unwrap();
+        sql_query(
+            r#"
+            CREATE TABLE transaction_depth_is_tracked_properly_on_commit_failure (
+                id SERIAL PRIMARY KEY,
+                class INTEGER NOT NULL
+            )
+        "#,
+        )
+        .execute(&conn)
+        .unwrap();
+
+        insert_into(serialization_example::table)
+            .values(&vec![
+                serialization_example::class.eq(1),
+                serialization_example::class.eq(2),
+            ])
+            .execute(&conn)
+            .unwrap();
+
+        let barrier = Arc::new(Barrier::new(2));
+        let threads = (1..3)
+            .map(|i| {
+                let barrier = barrier.clone();
+                thread::spawn(move || {
+                    use crate::connection::transaction_manager::AnsiTransactionManager;
+                    use crate::connection::transaction_manager::TransactionManager;
+                    let conn = crate::test_helpers::pg_connection_no_transaction();
+                    assert_eq!(0, <AnsiTransactionManager as TransactionManager<PgConnection>>::get_transaction_depth(&conn.transaction_manager));
+
+                    let result =
+                    conn.build_transaction().serializable().run(|| {
+                        let _ = serialization_example::table
+                            .filter(serialization_example::class.eq(i))
+                            .count()
+                            .execute(&conn)?;
+
+                        assert_eq!(1, <AnsiTransactionManager as TransactionManager<PgConnection>>::get_transaction_depth(&conn.transaction_manager));
```

Comment:
That check should be better done at the beginning of the transaction callback, otherwise it seems to be a bit out of place here, in between the different queries.